### PR TITLE
limit number of accumulated factors in running product to at most T

### PR DESCRIPTION
**Pull request by jan-ferdinand**
- Introduce register `ramp`, which is updated on executing instructions `read_mem` & `write_mem`.
- Explicitly treat clock jump differences of magnitude 1.
- Also, make running evaluation `reu` absorb first (not last) cell of every contiguous region. I believe relying on the existence of the padding region, which is by no means guaranteed, can lead to hard-to-catch bugs, both in the code and in soundness of the presented argument.

### CODE REVIEW

**Comment by aszepieniec** [comment]
> Introduce register `ramp` and instructions `set_ramp` & `get_ramp`.

Before having read the changes in detail, I wonder if we need to add the two instructions. The alternative would be to update the `ramp` register in the background, whenever `read_mem` or `write_mem` are executed.

**Comment by jan-ferdinand** [comment]
Makes perfect sense. Since neither `read_mem` nor `write_mem` changes the size of the stack, only one memory pointer is changed, namely, `ramp`. I seem to have gotten lost in the design possibilities and mixed things together.

**Review by aszepieniec — CHANGES_REQUESTED** [review]
The picture needs some work ;-)

File: tips/tip-0003/tip-0003.md [hunk 1]
```diff
@@ -13,32 +13,47 @@
...
+How to establish that a clock jump is directed forward (as opposed to directed backward, which would indicate malicious behavior)?
+One strategy is to show that the *difference*, *i.e.*, the next clock cycle minus the current clock cycle, is itself a clock cycle.
+Recall that the Processor Table's clock cycles run from 0 to $T-1$.
+Therefore, valid differences belong to $V = \lbrace 1, \dots, T-1 \rbrace \subseteq \mathbb{F}_p$ and invalid ones to $\mathbb{F}_p\setminus V$.
+As long as $T \ll p/2$, there is no overlap between these sets.
```
[the remark this answers is not in the document]
Comment (jan-ferdinand) [reply tips/tip-0003/tip-0003.md:27]: So the two sets are subsets of $\mathbb{Z}$, not of $\mathbb{F}_p$?

[the remark this answers is not in the document]
Comment (aszepieniec) [reply tips/tip-0003/tip-0003.md:27]: No, they are both subsets of $\mathbb{F}_p$. The key property we need is that these subsets are non-overlapping.

Maybe we should not call them *valid* and *invalid* clock jump differences, but *forward* and *backward* clock jump differences instead ...

[the remark this answers is not in the document]
Comment (jan-ferdinand) [reply tips/tip-0003/tip-0003.md:27]: I think that can help clear things up, yes. Additionally, I'll add type annotations of the form “ $\subseteq \mathbb{F}_p$ ”.

[the remark this answers is not in the document]
Comment (jan-ferdinand) [reply tips/tip-0003/tip-0003.md:27]: Are the set's boundaries correct? $(T - 1) + (p - T + 1) = p = 0$. I would expect the sets to be disjoint, thus the respective explicitly mentioned boundaries to have difference that's non-zero.

[the remark this answers is not in the document]
Comment (aszepieniec) [reply tips/tip-0003/tip-0003.md:27]: The correct boundaries are:

 - forward jumps: $\lbrace 1, \ldots, T \rbrace$
 - backward jumps: $\lbrace p-1-T, \ldots, p-1 \rbrace$.

Clock jump differences from outside these ranges simply cannot arise. Of the ones that can arise, the ones that lie in the backward jumps range are malicious.

[the remark this answers is not in the document]
Comment (jan-ferdinand) [reply tips/tip-0003/tip-0003.md:27]: The largest value for `clk` to ever be encountered in a Processor Table of length $T$ is $T-1$. I say:
 - forward jumps: $F = \\{ 1, \ldots, T-1 \\}$
 - backward jumps: $B = \\{-f \mid f \in F\\} = \\{-(T-1), \dots, -1 \\} = \\{ p-1-T, \ldots, p-1 \\}$.

**Pull request merged**